pkg/pool-weighted/test/foundry/WeightedMath.t.sol

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import { Test } from "forge-std/Test.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+
+import "../../contracts/WeightedMath.sol";
+
+contract WeightedMathTest is Test {
+    using FixedPoint for uint256;
+
+    function testJoinSwaps(
+        uint256[20] memory balancesFixed,
+        uint256[20] memory normalizedWeightsFixed,
+        uint256 arrayLength,
+        uint256 tokenIndex,
+        uint256 amountIn,
+        uint256 bptTotalSupply,
+        uint256 swapFeePercentage
+    ) external {
+        arrayLength = bound(arrayLength, 2, 20);
+        tokenIndex = bound(tokenIndex, 0, arrayLength - 1);
+
+        uint256[] memory balances = new uint256[](arrayLength);
+        for (uint256 i = 0; i < arrayLength; i++) {
+            balances[i] = bound(balancesFixed[i], 1e10, type(uint96).max);
+        }
+
+        uint256 denormalizedWeightSum;
+        for (uint256 i = 0; i < arrayLength; i++) {
+            normalizedWeightsFixed[i] = bound(normalizedWeightsFixed[i], 1, type(uint64).max);
+            denormalizedWeightSum += normalizedWeightsFixed[i];
+        }
+
+        uint256[] memory normalizedWeights = new uint256[](arrayLength);
+        uint256 normalizedWeightSum;
+        for (uint256 i = 0; i < arrayLength; i++) {
+            normalizedWeights[i] = normalizedWeightsFixed[i].divDown(denormalizedWeightSum);
+            vm.assume(normalizedWeights[i] >= WeightedMath._MIN_WEIGHT);
+            normalizedWeightSum += normalizedWeights[i];
+        }
+
+        // Note: Due to compression errors, this normalization property of weights may not always hold.
+        // This causes the two forms of join to produce slightly different outputs due to
+        // `WeightedMath._calcBptOutGivenExactTokenIn` assuming perfect normalization.
+        if (normalizedWeightSum < FixedPoint.ONE) {
+            normalizedWeights[arrayLength - 1] += FixedPoint.ONE - normalizedWeightSum;
+        }
+
+        amountIn = bound(amountIn, 0, balances[tokenIndex] / 100);
```

Comment:
Can this be relaxed?

Reply:
I've allowed it to be up to `balances[tokenIndex]`